File: cpp/myLeetCode/Dynamic_Programming/Basic/63. Unique Paths II.cpp
// 63. Unique Paths II
// https://leetcode.com/problems/unique-paths-ii/

#include <iostream>
#include <vector>

int uniquePathsWithObstacles(std::vector<std::vector<int>>& obstacleGrid) {
    int m = obstacleGrid.size();
    int n = obstacleGrid[0].size();

    // 如果起点或终点有障碍物，则路径数为 0
    if (obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1) {
        return 0;
    }

    std::vector<std::vector<long long>> dp(m, std::vector<long long>(n, 0));

    // 初始化边界条件
    dp[0][0] = 1;
    for (int i = 1; i < m; ++i) {
        if (obstacleGrid[i][0] == 0) {
            dp[i][0] = dp[i-1][0];
        }
    }
    for (int j = 1; j < n; ++j) {
        if (obstacleGrid[0][j] == 0) {
            dp[0][j] = dp[0][j-1];
        }
    }

    // 动态规划计算路径数
    for (int i = 1; i < m; ++i) {
        for (int j = 1; j < n; ++j) {
            if (obstacleGrid[i][j] == 0) {
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
            }
        }
    }

    return dp[m-1][n-1];
}

int main() {
    std::vector<std::vector<std::vector<int>>> testCases = {
        {{0, 0, 0}, {0, 1, 0}, {0, 0, 0}},
        {{0, 1}, {0, 0}}
    };
    for (auto& testCase: testCases) {
        std::cout << "Number of unique paths: " << uniquePathsWithObstacles(testCase) << std::endl;
    }
    
    return 0;
}

File: C++/代码随想录/Dynamic_Programming/Basic/63. Unique Paths II.py
def uniquePathsWithObstacles(obstacleGrid):
    if not obstacleGrid or obstacleGrid[0][0] == 1:
        return 0

    m, n = len(obstacleGrid), len(obstacleGrid[0])
    dp = [[0] * n for _ in range(m)]
    dp[0][0] = 1

    for i in range(1, m):
        if obstacleGrid[i][0] == 0:
            dp[i][0] = dp[i - 1][0]

    for j in range(1, n):
        if obstacleGrid[0][j] == 0:
            dp[0][j] = dp[0][j - 1]

    for i in range(1, m):
        for j in range(1, n):
            if obstacleGrid[i][j] == 0:
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

    return dp[m - 1][n - 1]

# 测试用例
print(uniquePathsWithObstacles([[0,0,0],[0,1,0],[0,0,0]]))  # 应该输出 2
print(uniquePathsWithObstacles([[0,1],[0,0]]))  # 应该输出 1
